### PR TITLE
Write path error handling cleanup

### LINKED ISSUE #3096: Concurrent writes may get Status::OK while it actually gets IOError on WAL write
Concurrent writes may get Status::OK while it actually gets IOError on WAL write. This happens when multiple writes form a write batch group, and the leader get an IOError while writing to WAL. The leader failed to pass the error to followers in the group, and the followers end up return Status::OK() while actually writing nothing. The bug only affect writes in a batch group. Future writes after the batch group will correctly return immediately with the IOError.

Affects RocksDB >= 4.6.1 and <= 5.8.x since this commit: 6f71d3b68b011569dfaf5c406a15319e4d66ca76

### PR DESCRIPTION
**Pull request by yiwu-arbug**
Summary:
Fixing the bug where concurrent writes may get Status::OK while it actually gets IOError on WAL write. This happens when multiple writes form a write batch group, and the leader get an IOError while writing to WAL. The leader failed to pass the error to followers in the group, and the followers end up return Status::OK() while actually writing nothing. The bug only affect writes in a batch group. Future writes after the batch group will correctly return immediately with the IOError.

Closes #3096 

Test Plan:
See the new test.

### CODE REVIEW

File: db/db_impl_write.cc [hunk 1]
```diff
@@ -329,7 +330,7 @@ Status DBImpl::WriteImpl(const WriteOptions& write_options,
       versions_->SetLastSequence(last_sequence);
     }
     MemTableInsertStatusCheck(w.status);
-    write_thread_.ExitAsBatchGroupLeader(write_group, w.status);
+    write_thread_.ExitAsBatchGroupLeader(write_group, status);
```
Comment (maysamyabandeh) [inline db/db_impl_write.cc]: This needs to be updated since status is not just about memtable write status:
> Status status;            // status of memtable inserter

Comment (maysamyabandeh) [reply db/db_impl_write.cc]: Here we assume the status is ok when the follower completes. Is that in contradiction with assigning status to non ok when there is a wal error?
```
518 void WriteThread::ExitAsBatchGroupFollower(Writer* w) {
519   auto* write_group = w->write_group;
520
521   assert(w->state == STATE_PARALLEL_MEMTABLE_WRITER);
522   assert(write_group->status.ok());
523   ExitAsBatchGroupLeader(*write_group, write_group->status);
524   assert(w->status.ok());
525   assert(w->state == STATE_COMPLETED);
526   SetState(write_group->leader, STATE_COMPLETED);
527 }
```

Comment (maysamyabandeh) [reply db/db_impl_write.cc]: Here we replace writer status with wal status. What if wal succeeded but the memtable write had failed?

Comment (maysamyabandeh) [reply db/db_impl_write.cc]: How about other usages of ExitAsBatchGroupLeader in PipelinedWriteImpl and WriteImplWALOnly?

**Comment by yiwu-arbug** [comment]
I'm doing more changes around the fix. Here's the change:
* Remove local variable `status` in `WriteImpl` family methods and use `w.status` instead when needed. This is to avoid confusing between when to use which status variable.
* Remove `status` param from `WriteThread::ExitAsBatchGroupLeader` for the same reason. The method now update all writers in the group with `write_group.status`.
* There are now two places we update `write_group.status`:
    - In `CompleteParallelMemTableWriter` where parallel memtable writer report its status back to the write group.
    - The leader will report status to the write group before it calls `WriteThread::ExitAsBatchGroupLeader`

File: db/db_impl_write.cc [hunk 2]
```diff
@@ -471,7 +467,7 @@ Status DBImpl::PipelinedWriteImpl(const WriteOptions& write_options,
         write_options.ignore_missing_column_families, 0 /*log_number*/, this,
         true /*concurrent_memtable_writes*/);
     if (write_thread_.CompleteParallelMemTableWriter(&w)) {
-      MemTableInsertStatusCheck(w.status);
+      SetBackgroundErrorIfNeeded(w.write_group->status);
```
Comment (maysamyabandeh) [inline db/db_impl_write.cc]: What happens to w.status? Previously it could set the background error but now it is simply returned to the user.

Comment (maysamyabandeh) [reply db/db_impl_write.cc]: Do we need to set the write_group status in PipelinedWriteImpl the way we do for WriteImpl?

File: db/db_impl_write.cc [hunk 3]
```diff
@@ -563,33 +559,41 @@ Status DBImpl::WriteImplWALOnly(const WriteOptions& write_options,
...
+    SetBackgroundErrorIfNeeded(w.status);
   }
+  // Propagate error, if any, to the whole write group
+  write_group.UpdateStatus(w.status);
+  nonmem_write_thread_.ExitAsBatchGroupLeader(write_group);
   if (seq_used != nullptr) {
     *seq_used = w.sequence;
   }
-  return status;
+  return w.FinalStatus();
```
Comment (maysamyabandeh) [inline db/db_impl_write.cc]: Previously w.FinalStatus would be returned if the status (w.status now) was Ok otherwise status (w.status) would be returned. But now we always return w.FinalStatus regardless. Can you explain why this change is ok?

File: db/db_impl_write.cc [hunk 4]
```diff
@@ -318,21 +316,20 @@ Status DBImpl::WriteImpl(const WriteOptions& write_options,
   }
   PERF_TIMER_START(write_pre_and_post_process_time);
 
-  if (!w.CallbackFailed()) {
-    WriteCallbackStatusCheck(status);
```
[the remark this answers is not in the document]
Comment (yiwu-arbug) [reply db/db_impl_write.cc]: Make sense. Let me put back the bg error check

File: db/db_impl_write.cc [hunk 5]
```diff
@@ -344,17 +341,15 @@ Status DBImpl::WriteImpl(const WriteOptions& write_options,
     should_exit_batch_group = write_thread_.CompleteParallelMemTableWriter(&w);
   }
   if (should_exit_batch_group) {
-    if (status.ok()) {
```
[the remark this answers is not in the document]
Comment (yiwu-arbug) [reply db/db_impl_write.cc]: To be consistent with this line where if the leader exit the batch group, it doesn't check status. I think we probably should check `if (status.ok() || !db_options.paranoid_check)` instead? Suggestion is welcome.
https://github.com/facebook/rocksdb/blob/master/db/db_impl_write.cc#L348

File: db/db_impl_write.cc (hunk 2 above)
Comment (yiwu-arbug) [reply db/db_impl_write.cc]: yeah, I updated CompleteParallelMemTableWriter to set write_group status on CompleteParallelMemTableWriter, same as in PiplinedWriteImpl.

File: db/db_impl_write.cc (hunk 3 above)
Comment (yiwu-arbug) [reply db/db_impl_write.cc]: With the patch `w.status` is the same as previous `status`, which contain preprocess error, WAL error and memtable error. `w.FinalStatus` will return only be different from `w.status` when `w.status.ok() && !callback_status.ok()`. So the change here is  equivalent to previous version.

File: db/db_impl_write.cc [hunk 6]
```diff
@@ -607,27 +611,6 @@ void DBImpl::WriteCallbackStatusCheck(const Status& status) {
   }
 }
 
-void DBImpl::MemTableInsertStatusCheck(const Status& status) {
-  // A non-OK status here indicates that the state implied by the
-  // WAL has diverged from the in-memory state.  This could be
-  // because of a corrupt write_batch (very bad), or because the
-  // client specified an invalid column family and didn't specify
-  // ignore_missing_column_families.
-  if (!status.ok()) {
```
[the remark this answers is not in the document]
Comment (yiwu-arbug) [reply db/db_impl_write.cc]: According to description for `paranoid_checks` we should only stop writes if `paranoid_checks == true`.
https://github.com/facebook/rocksdb/blob/master/include/rocksdb/options.h#L338-L344

The memtable status check to set bg_error was added in this commit, but it didn't check if `paranoid_checks` is set.
6ce42dd0750def33aaa632d064ba0a097ac05b14

It seems to me for the behavior to be consistent we should check if `paranoid_checks` is set for memtable error as well. It is okay to me to keep the existing behavior, but it would make the logic a bit weird. Opinions?

**Comment by yiwu-arbug** [comment]
@maysamyabandeh kindly ping

**Comment by sagar0** [comment]
@yiwu-arbug I guess this can be closed and discarded now that #3201 landed?

**Comment by yiwu-arbug** [comment]
@sagar0 This one include multiple fixes that #3201 don't include. I'm keeping the PR and will get back to it once I have time.

**Comment by maysamyabandeh** [comment]
@anand1976 Given your recent work on error handling, would you be interested to evaluate this PR? I am not sure how much of it is already covered by other works you did in the past year.

File: db/db_impl_write.cc [hunk 7]
```diff
@@ -563,33 +559,41 @@ Status DBImpl::WriteImplWALOnly(const WriteOptions& write_options,
...
+  // 1. Errors from SwitchWAL or SwitchMemTable
+  // 2. Error on WAL write
+  // 3. Error on memtable write, which can be
+  //   3a) A corrupted write_batch (very bad), or
+  //   3b) client specified an invalid column family and didn't specify
+  //       ignore_mising_column_families
+  //
+  // The following should be seen as background error:
+  //   * IsBusy which can come from transactions.
+  //   * IsIncomplete can come from DelayWrite.
```
Comment (anand1976) [inline db/db_impl_write.cc]: The code below needs to be changed to call ```error_handler_.SetBGError()```.

**Comment by anand1976** [comment]
@yiwu-arbug Thanks. I think this bug still exists and needs to be fixed. I'll commandeer this PR.

**Comment by anand1976** [comment]
On further examination, it looks like https://github.com/facebook/rocksdb/commit/3cf562be31a5876a00caff6eb638049bcad5ee9b already fixed this issue in a simpler way. Closing this PR as its no longer needed.
